### DISCUSSION TITLE
Evaluating Vulkan graphics -- testers needed

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#187] [General]
Greetings all,

Andrew @awolven has implemented a first pass at Vulkan graphics, and we need some people to test it out.

Documentation is below, and here are the links:

Andrew's kons-9 with krma:  https://github.com/awolven/kons-9/tree/krma

krma:  https://github.com/awolven/krma

cl-vulkan:  https://github.com/awolven/cl-vulkan

Please let us know how your tests go.

[krma documentation.pdf](https://github.com/kaveh808/kons-9/files/10024972/krma.documentation.pdf)

**Comment by awolven** [comment]
Some things to note:

MacOS does not have the ability to set :line-thickness.

AMD driver on Windows, and possibly Linux has a bug which only sometimes
shows points.

When you make posts about bugs/performance, include:

Your OS, and if Linux X11 or Wayland, your video card/driver i.e. NVidia/
AMD/ Intel/ M1.

Triangle strip and triangle list primitives have not been tested, but
polygons/polylines/lines/circles/rectangles/spheres/text should be ok.

The 3d-bmfont which text in krma depends on, has a problem with pathnames,
so you may need to find "acache.png" and "acache.json" and stick them in
krma/.  They don't get generated until the first call to "main" or
krma::run on your system.

I appreciate all the help.

Thanks,
Andrew

**Comment by awolven** [comment]
Another, thing,

I modified kons-9::run to run in the current thread because of a bug on
Windows, for MacOS you either need to change it back or launch it from your
*inferior-lisp* buffer.

-AKW

**Comment by kaveh808** [comment]
@nsmoker @JMC-design @foretspaisibles Please have a look if you have the time.

**Comment by lukego** [comment]
~How come `krma` references its dependencies with Git submodules? Is that just for convenience or is it really important to pin those precise versions of the dependencies?~

Forget I asked: Clearly these versions are required instead of the mainlines.

**Comment by lukego** [comment]
@awolven I'm able to build and start the `krma` branch!

I'm hitting an error at startup though:

```
* (kons-9::run)
WARNING:
   Trying to create vulkan instance with VK_LAYER_LUNARG_STANDARD_VALIDATION failed, falling back...

debugger invoked on a FILE-DOES-NOT-EXIST in thread
#<THREAD "main thread" RUNNING {1001788113}>:
  The file
  #P"/nix/store/bjqfh04s1xb2r0vgfqf0bj5bf67igx6s-krma-20221115-8ba1252/standard-2d.vert.spv"
  does not exist:
    No such file or directory
```

I'm out of time right now so I can't dig into where the .spv files are supposed to come from. I guess those are the vulkan shader subroutines.

Screenshot: https://imgur.com/a/lnhW3Ck

**Comment by awolven** [comment]
The directions on how to compile the shaders are in the documentation.

**Comment by awolven** [comment]
I don't want to put binaries in the repo at this time.  When vulkan gets
completely integrated with kons-9, we can distribute the spirv compiler per
the license in the SDK as part of some kind of installer or something.  But
for now, to avoid spamming the repos, [re]compiling the shaders is a
separate step.

**Comment by awolven** [comment]
In the interim, for testers, maybe I can make a separate repo which is just
the latest shader binaries, and make it a submodule.

**Comment by awolven** [comment]
Ok, I did that, you can now pull krma and do a git submodule update --init
again and you should have the spv files.

**Comment by awolven** [comment]
This should mean that you don't necessarily have to download the Vulkan SDK
on Windows or Linux (unless you want the vulkan validation layers), but on
MacOS you're going to need the SDK to get the MoltenVK client driver.

**Comment by lukego** [comment]
Thanks for pointing out the docs @awolven. I didn't realize there were installation instructions in the PDF file. I'll revisit this soon and see what I can get working.

**Comment by lukego** [comment]
Thanks for the update. I pulled the new submodule and now I'm seeing a font-related error instead. Hard-coded paths like `/usr/share/fonts/` don't work on my system so I'll need to find which code is looking there and find an alternative e.g. environment variable where I can point it.

```
* (kons-9::run)
WARNING:
   Trying to create vulkan instance with VK_LAYER_LUNARG_STANDARD_VALIDATION failed, falling back...

debugger invoked on a FILE-DOES-NOT-EXIST in thread
#<THREAD "main thread" RUNNING {10017A0113}>:
  The file #P"/usr/share/fonts/liberation-mono/LiberationMono-Regular.ttf" does
  not exist:
    No such file or directory

Type HELP for debugger help, or (SB-EXT:EXIT) to exit from SBCL.

restarts (invokable by number or by possibly-abbreviated name):
  0: [CREATE   ] Reopen with :if-does-not-exist :create
  1: [CONTINUE ] Retry opening.
  2: [USE-VALUE] Try opening a different file.
  3: [ABORT    ] Exit debugger, returning to top level.

(SB-IMPL::FILE-PERROR #P"/usr/share/fonts/liberation-mono/LiberationMono-Regular.ttf" 2 FILE-DOES-NOT-EXIST)
0] backtrace

Backtrace for: #<SB-THREAD:THREAD "main thread" RUNNING {10017A0113}>
0: (SB-IMPL::FILE-PERROR #P"/usr/share/fonts/liberation-mono/LiberationMono-Regular.ttf" 2 FILE-DOES-NOT-EXIST)
1: (SB-IMPL::%OPEN-ERROR #P"/usr/share/fonts/liberation-mono/LiberationMono-Regular.ttf" 2 :IGNORE-THIS-ARG :ERROR)
2: (OPEN "/usr/share/fonts/liberation-mono/LiberationMono-Regular.ttf" :DIRECTION :INPUT :ELEMENT-TYPE (UNSIGNED-BYTE 8) :IF-EXISTS NIL :IF-DOES-NOT-EXIST NIL :EXTERNAL-FORMAT :DEFAULT :CLASS SB-SYS:FD-STREAM)
3: (ZPB-TTF::OPEN-FONT-LOADER-FROM-FILE "/usr/share/fonts/liberation-mono/LiberationMono-Regular.ttf" :COLLECTION-INDEX 0)
4: (SDF:MAKE-ATLAS "/usr/share/fonts/liberation-mono/LiberationMono-Regular.ttf" 16 :SPREAD 8 :STRING NIL :WIDTH :AUTO :HEIGHT :AUTO :MODE :MSDF+A :AUTO-SIZE-GRANULARITY-X 1 :AUTO-SIZE-GRANULARITY-Y 1 :OPTIMIZE-PACK NIL :TRIM NIL :VERBOSE NIL :EXPAND-MODE #<unused argument> :SCALE #<unused argument>)
5: (SDF-BMFONT:CREATE-BMFONT "/usr/share/fonts/liberation-mono/LiberationMono-Regular.ttf" "acache.json" :SIZE 16 :MODE :MSDF+A :TYPE :JSON :SPREAD 8)
6: ((:METHOD VK:MAIN (KONS-9::KONS-9)) #<KONS-9::KONS-9 {10045B28D3}>) [fast-method]
7: (SB-INT:SIMPLE-EVAL-IN-LEXENV (KONS-9:RUN) #<NULL-LEXENV>)
8: (EVAL (KONS-9:RUN))
9: (INTERACTIVE-EVAL (KONS-9:RUN) :EVAL NIL)
10: (SB-IMPL::REPL-FUN NIL)
11: ((LAMBDA NIL :IN SB-IMPL::TOPLEVEL-REPL))
12: (SB-IMPL::%WITH-REBOUND-IO-SYNTAX #<FUNCTION (LAMBDA NIL :IN SB-IMPL::TOPLEVEL-REPL) {1002BA40CB}>)
13: (SB-IMPL::TOPLEVEL-REPL NIL)
14: (SB-IMPL::TOPLEVEL-INIT)
15: ((FLET SB-UNIX::BODY :IN SB-IMPL::START-LISP))
16: ((FLET "WITHOUT-INTERRUPTS-BODY-3" :IN SB-IMPL::START-LISP))
17: (SB-IMPL::%START-LISP)
```

**Comment by awolven** [comment]
Just try using any .ttf you like.  The code is in
kons-9/src/kernel/main.lisp in the main method.

**Comment by lukego** [comment]
@awolven Hey it's working now! Feels very smooth and the text in the menus even works now (whereas main branch menu text has never worked for me due to #125.)

I'll have a more proper look tomorrow but amazing fast work you have done here.

I did have to substitute in a TTF font and copy the `acache.json` and `acache.png` files into the `krma/` directory as mentioned above ^.

**Comment by lukego** [comment]
Pic or it didn't happen:

![image](https://user-images.githubusercontent.com/13791/202796317-198792bf-e5c8-4c60-a503-ca9baf1d9003.png)

**Comment by awolven** [comment]
Try evaluating some of the forms in test/demo-kernel.lisp, the polyhedrons
should work.

**Comment by lukego** [comment]
I'll kick the tires better tomorrow. It's bedtime here now. Thanks for the help!

**Comment by lukego** [comment]
@awolven In `test/demo-kernel.lisp` most stuff is working but some is not. What constitutes useful feedback?

It's dead sexy with the shading:

![Screenshot from 2022-11-19 09-50-11](https://user-images.githubusercontent.com/13791/202843112-a1f5643e-b205-4cde-91cc-91c686817326.png)

**Comment by awolven** [comment]
The shading is not quite right.  The light is supposed to be coming from
somewhere behind the eye position in that picture, but the octahedron and
dodecahedron are shaded wrong.

Please list your OS,

Printed representation of:
vk::(physical-device (default-logical-device *app*))

Wayland or X11

(princ vk::*present-mode*)

And if you're brave, for the demos that don't work, step through the draw
routines for the classes in question, and make sure it's not calling opengl.

Also, if you don't see vertex points, I'm hoping that's because you've got
an AMD card and then it isn't such a mystery.

AKW

**Reply by lukego** [reply]
Here's a video of walking through the kernel demo: https://www.youtube.com/watch?v=6Gr7tXIkyf8. It will be in 4K once Youtube finishes processing it.

```
OS: NixOS Linux
CPU: AMD Ryzen 7 3700X
GPU: Nvidia RTX 3080

X11

vk::(physical-device (default-logical-device krma::*app*))
  => #<GPU NVIDIA GeForce RTX 3080>

(princ vk::*present-mode*)
  => 0
```

How do I check mesa vs native driver?

**Comment by awolven** [comment]
Oh yeah,

List mesa or native driver.

**Comment by awolven** [comment]
Try out the retained-mode API

**Reply by lukego** [reply]
> Try out the retained-mode API

How?

**Comment by awolven** [comment]
Type this at a REPL:

(krma:add-text "The quick brown fox jumps over the lazy dog." 100 200)

It's all in the documentation.

**Comment by awolven** [comment]
Since the background is white, you better do:

(krma:add-text "The quick brown fox jumps over the lazy dog." 100 200
:color #x0000ffff)

**Reply by lukego** [reply]
Oh, neat! Yes if I paste in that code then it draws on the screen and is preserved when I manipulate the kons-9 UI beyhind :+1: 

![image](https://user-images.githubusercontent.com/13791/202869359-9f3eef4e-af35-4e7d-b7be-891c33e640a5.png)

**Comment by awolven** [comment]
try (krma:delete-group (krma:application-scene vk:*app*) :default)

**Reply by lukego** [reply]
The text disappeared and the kons-9 UI is still there and working.

**Comment by cmack** [comment]
I was able to get this running on a GUIX system relatively easily once I tackled the dependency list. My experience following the documentation:
1. The `git submodules` command didn't install `cl-vulkan`, but installing using similar technique to `krma` worked
2. If anyone else is in the same boat, this is how I initialized my dev environment. The wayland dependencies might be unnecessary -- part of my recent habits elsewhere.
```sh
# environment with all dependencies
guix shell sbcl wayland wayland-protocols mesa glslang vulkan-loader vulkan-headers vulkan-tools glfw font-liberation freeglut
# start emacs in a way that SLIME can find the dev libs
LD_LIBRARY_PATH="$GUIX_ENVIRONMENT/lib" emacs& 
```
3. I had to modify the `compile.sh` to find my `bash` and `env` locations but it ran ok
4. As already mentioned above, I also had to change the hardcoded path for LiberationMono-Regular 
5. I used `(bt:make-thread #'kons-9:run)` -- seems ok.
![20221120_15h54m26s_grim](https://user-images.githubusercontent.com/723154/202928763-f4b184eb-a880-4621-9c49-017a80280380.png)

**Comment by awolven** [comment]
These are the things I plan on doing to krma and the kons-9 krma port:


   - repair create-bmfont to not "lose the cache file"
   - it's putting a bogus path in the .json file
      - make sure font engine works for all .ttf files
   - doesn't work for certain dejavu fonts for example
      - test for correct operation in long-running processes
   - i.e. no memory leaks/latency issues
      - get all platforms respecting system resources
   - try to get 60 fps on non-high-performance displays
      - try to get 120 fps on high-performance displays
      - make moving glfw window between monitors work well
      - try to use present-mode-fifo when the first two bullets can be
      satisfied
      - note: windows on a non-high-performance monitor with an APU gets 60
      fps and 1.4% cpu which is nice (in present mode immediate)
      - make draw-list compaction work for retained mode (and efficiently)
   - get rid of reinstance-primitive for now
      - work on polygon-based thick lines for 2d and cylinder based thick
   lines for 3d so all platforms will look the same, and nice.
   - spherical or dodecahedron points
      - get lighting working better
   - package "roboto" font with krma, for a default font
   - make cl-vulkan a submodule of krma and krma a submodule of kons-9
   - get cl-metal going to at least allow compute shaders on Apple
   - make repo for objc bridge
      - test all graphics primitives (i.e. triangle list/triangle
   strip/flat-shaded polygons)
   - get automagic doc-string to markdown documenter running
   - edit documentation to make more readable and understandable


So testers needn't open issues on those points.  But I still need people to
evaluate the architecture, to make sure it is sane, usable for kons-9, and
hopefully even fun.  You don't even have to build the kons-9/krma branch to
do this.  Reading the documentation in partial would be good enough.  The
next level would be to try drawing stuff in the kons-9 krma branch, and
after that the next level is to review the source.  I don't want to spend a
lot of time polishing krma, until I can at least be assured it is sane and
performant, which is why I gave it to Kaveh in the first place.

-AKW

**Comment by kaveh808** [comment]
Sounds great, Andrew. I'll make some time to read the docs.

We could consider progressing in increments: first replace the existing OpenGL functionality, and later convert to more optimized Vulkan calls.

**Comment by awolven** [comment]
It's probably important to keep in mind, that if/when selection is
implemented by means of using an auxiliary frame buffer, the API will
likely change to have that object id be passed in to the group or primitive
creation, where it will pass through the shader and ultimately back to the
user as a set of "hovered" objects.  But what I'm saying is that there
could be a minor api change.

**Comment by kaveh808** [comment]
No worries about API changes. In my mind kons-9 is in an early design state and anything we need can be changed.
